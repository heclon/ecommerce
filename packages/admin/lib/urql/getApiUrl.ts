const getApiUrl = (): string => {
  // In the browser we just use a relative URL and everything works perfectly
  if (process.browser) return `/api`

  // Infer the deploy URL if we're in production
  const PROVIDER_URL = process.env.ADMIN_DOMAIN

  if (PROVIDER_URL) {
    // We replace https:// from the URL if it exists and add it ourselves always at the beginning as the above environment variables are not guaranteed to include it
    return `https://${PROVIDER_URL.replace(/^https?:\/\//, '')}/api`
  }

  // Finally, fallback to hard-coded URL in case nothing else works
  if (process.env.NODE_ENV === `development`) return `http://localhost:3001/api`

  return 'https://admin.ecommerce.com.au/api'
}

export default getApiUrl
