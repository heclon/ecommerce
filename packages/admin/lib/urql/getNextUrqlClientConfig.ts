import { cacheExchange } from '@urql/exchange-graphcache'
import { multipartFetchExchange } from '@urql/exchange-multipart-fetch'
import { dedupExchange, errorExchange } from 'urql'
import { devtoolsExchange } from '@urql/devtools'
import { NextUrqlClientConfig } from 'next-urql'
import getApiUrl from './getApiUrl'
import uniqBy from 'lodash/uniqBy'
import toast from 'react-hot-toast'

export const getNextUrqlClientConfig: NextUrqlClientConfig = (ssrExchange) => ({
  exchanges: [
    devtoolsExchange,
    errorExchange({
      onError: (error) => {
        if (error.networkError) {
          console.error(`[HH-Admin: Network error]: ${error.message} ${error.stack}`)
        } else {
          const errors = uniqBy(error.graphQLErrors, (error) => error.message)
          errors.forEach((error) => {
            console.error(`[HH-Admin: GraphQL error]: Message: ${error.message}, Path: ${error.path}`)
          })
        }
        toast.error(
          `Error occurred, our team has been notified and is looking into it. \nSomething urgent? \nPlease contact the engineering team
    in the #tech-issues Slack channel`
        )
      },
    }),
    dedupExchange,
    cacheExchange(),
    ssrExchange,
    multipartFetchExchange,
  ],
  fetchOptions: {
    credentials: 'include',
  },
  requestPolicy: `cache-and-network`,
  url: getApiUrl(),
})

export default getNextUrqlClientConfig
