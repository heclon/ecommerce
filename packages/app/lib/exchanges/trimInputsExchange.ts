import { Exchange, Operation } from 'urql'
import { pipe, tap } from 'wonka'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const trimInputs = (input: any): any => {
  if (typeof input === 'string') {
    return input.trim()
  }
  if (Array.isArray(input)) {
    return input.map((value) => trimInputs(value))
  }
  if (input instanceof Set) {
    return new Set([...input].map((value) => trimInputs(value)))
  }
  if (input instanceof Map) {
    return [...input.entries()].reduce((map, [key, value]) => {
      map.set(key, trimInputs(value))
      return map
    }, new Map())
  }
  if (input != null && typeof input == 'object' && toString.call(input) == '[object Object]') {
    return Object.entries(input).reduce(
      (obj, [key, value]) => ({
        ...obj,
        [key]: trimInputs(value),
      }),
      {}
    )
  }
  // do nothing.
  return input
}

export const trimInput = (operation: Operation) => {
  operation.variables = trimInputs(operation.variables)
  return operation
}

export const trimInputsExchange: Exchange = ({ forward }) => {
  return (ops$) => {
    return pipe(ops$, tap(trimInput), forward)
  }
}
