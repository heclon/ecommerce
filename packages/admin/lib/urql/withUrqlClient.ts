import { withUrqlClient as withUrqlClientWrapper } from 'next-urql'
import getNextUrqlClientConfig from './getNextUrqlClientConfig'
import { NextPage } from 'next'

const withUrqlClient = <T>(component: NextPage<T>) => {
  return withUrqlClientWrapper(getNextUrqlClientConfig)(component)
}

export default withUrqlClient
