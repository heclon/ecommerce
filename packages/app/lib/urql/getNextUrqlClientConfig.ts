import { trimInputsExchange } from '../exchanges/trimInputsExchange'
import getApiUrl from './getApiUrl'
import { NextUrqlClientConfig } from 'next-urql'
import { devtoolsExchange } from '@urql/devtools'
import { errorExchange, dedupExchange } from 'urql'
import { multipartFetchExchange } from '@urql/exchange-multipart-fetch'
import { cacheExchange } from '@urql/exchange-graphcache'
import uniqBy from 'lodash/uniqBy'
import { toast } from 'react-hot-toast'

const getNextUrqlClientConfig: NextUrqlClientConfig = (ssrExchange) => ({
  exchanges: [
    devtoolsExchange,
    errorExchange({
      onError: (error) => {
        if (error.networkError) {
          console.error(`[HH-App: Network error]: ${error.message} ${error.stack}`)
        } else {
          const errors = uniqBy(error.graphQLErrors, (error) => error.message)
          errors.forEach((error) => {
            console.error(`[HH-App: GraphQL error]: Message: ${error.message}, Path: ${error.path}`)
          })
        }

        if (error.message.includes('User email not verified')) {
          return toast.error('Please verify your email.')
        }

        toast.error(
          `Error occurred, our team has been notified and is looking into it. \nSomething urgent? \nPlease contact us on providers@ecommerce.com.au`
        )
      },
    }),
    trimInputsExchange,
    dedupExchange,
    cacheExchange(),
    ssrExchange,
    multipartFetchExchange,
  ],
  fetchOptions: {
    credentials: 'include',
  },
  requestPolicy: `cache-and-network`,
  url: getApiUrl(),
})

export default getNextUrqlClientConfig
