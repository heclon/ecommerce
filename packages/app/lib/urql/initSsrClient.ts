import { cacheExchange } from '@urql/exchange-graphcache'
import { errorExchange, dedupExchange, ssrExchange } from 'urql'
import { devtoolsExchange } from '@urql/devtools'
import { initUrqlClient, SSRExchange } from 'next-urql'
import { GetServerSidePropsContext } from 'next'
import { getSession } from '@auth0/nextjs-auth0'
import { executeExchange } from '@urql/exchange-execute'
import { Client } from '@urql/core'
import getApiUrl from './getApiUrl'
import { createContext } from '../../server/context'
import schema from '../../server/schema'
import { trimInputsExchange } from '../exchanges/trimInputsExchange'
import uniqBy from 'lodash/uniqBy'

const initSsrClient: (ctx: GetServerSidePropsContext) => Promise<[Client | null, SSRExchange]> = async (
  ctx: GetServerSidePropsContext
) => {
  const session = await getSession(ctx.req, ctx.res)
  const ssrCache = ssrExchange({ isClient: false })
  const client = initUrqlClient(
    {
      url: getApiUrl(), // not needed without `fetchExchange`
      exchanges: [
        devtoolsExchange,
        errorExchange({
          onError: (error) => {
            if (error.networkError) {
              console.error(`[HH-App: Network error]: ${error.message} ${error.stack}`)
            } else {
              const errors = uniqBy(error.graphQLErrors, (error) => error.message)
              errors.forEach((error) => {
                console.error(`[HH-App: GraphQL error]: Message: ${error.message}, Path: ${error.path}`)
              })
            }
          },
        }),
        trimInputsExchange,
        dedupExchange,
        cacheExchange(),
        ssrCache,
        executeExchange({
          schema,
          context: () => {
            return createContext({ session })
          },
        }),
      ],
    },
    false
  )
  return [client, ssrCache]
}

export default initSsrClient
