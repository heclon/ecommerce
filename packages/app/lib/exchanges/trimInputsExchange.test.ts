import { trimInput } from './trimInputsExchange'
import { Operation } from 'urql'

describe('trimInputsExchange', () => {
  it('works', () => {
    const operation = {
      variables: {
        testWhiteSpaces: '    String with whitespaces before and after    ',
        array: [
          {
            id: '1',
            value: '   whitespaces in Arrays  ',
          },
          {
            id: '2',
            value: '   another value with whitespaces   ',
          },
        ],
        null: null,
        set: new Set(['     whitespaces in Sets     ', '    another string to fix  ']),
        /*eslint-disable @typescript-eslint/no-explicit-any*/
        map: new Map<any, any>([
          ['x', '   y   '],
          ['z', '   whitespaces in Maps   '],
        ]),
      },
    } as unknown as Operation
    const operationWithTrimmedVars = trimInput(operation)
    expect(operationWithTrimmedVars).toMatchSnapshot()
  })
})
