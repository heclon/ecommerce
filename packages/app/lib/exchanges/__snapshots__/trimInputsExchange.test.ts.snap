// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trimInputsExchange works 1`] = `
Object {
  "variables": Object {
    "array": Array [
      Object {
        "id": "1",
        "value": "whitespaces in Arrays",
      },
      Object {
        "id": "2",
        "value": "another value with whitespaces",
      },
    ],
    "map": Map {
      "x" => "y",
      "z" => "whitespaces in Maps",
    },
    "null": null,
    "set": Set {
      "whitespaces in Sets",
      "another string to fix",
    },
    "testWhiteSpaces": "String with whitespaces before and after",
  },
}
`;
